### DISCUSSION TITLE
Agent with a flow chart image as a backstory

### DISCUSSION THREAD

**Discussion by gcleaves** [crewAIInc/crewAI#1520] [Q&A]
I have a complicated workflow which is best represented by a flow chart. GPT-4o has proven to understand the image well. Is it possible to somehow provide this flowchart knowledge to an agent?

**Comment by Hassanmahmood4** [comment]
Yes, although the approach depends on how "dynamic" you need the flowchart understanding to be.

If GPT-4o already understands the flowchart well, one straightforward approach is to include the flowchart image as part of the agent's context and let a multimodal model interpret it during execution.

However, if the flowchart represents relatively stable business logic, I would consider converting it into a machine-readable intermediate representation instead of repeatedly passing the image.

For example:

1. Use GPT-4o once to transform the flowchart into structured data:

```json
{
  "nodes": [
    {"id": "A", "type": "decision", "description": "..."},
    {"id": "B", "type": "action", "description": "..."}
  ],
  "edges": [
    {"from": "A", "to": "B", "condition": "..."}
  ]
}
```

2. Store that representation (JSON/YAML/database).

3. Give the agent access to it through a tool or retrieval step.

This has a few advantages:

- The agent can reason over the workflow deterministically.
- You avoid repeatedly paying for image understanding.
- The workflow becomes versionable and editable.
- Other agents can consume the same representation.

If the flowchart changes frequently and visual details matter, then keeping the image and using a multimodal model (e.g., GPT-4o) at runtime is perfectly reasonable.

Out of curiosity, is your goal for the agent to:

- simply **understand and execute** the flowchart,
- **modify/update** the flowchart,
- or **navigate it step-by-step as a state machine**?

The answer would influence whether I'd recommend multimodal prompting, retrieval over a structured representation, or implementing the workflow as an explicit graph/state machine.
